python/ray/experimental/dag/dag_node.py

```diff
@@ -0,0 +1,217 @@
+import ray
+from ray.experimental.dag.py_obj_scanner import _PyObjScanner
+from ray.experimental.dag.format_utils import (
+    get_args_lines,
+    get_kwargs_lines,
+    get_options_lines,
+    get_indentation,
+)
+import ray.experimental.dag as dag
+
+from typing import (
+    Optional,
+    Union,
+    List,
+    Tuple,
+    Dict,
+    Any,
+    TypeVar,
+    Callable,
+    Set,
+)
+import uuid
+
+T = TypeVar("T")
+
+
+class DAGNode:
+    """Abstract class for a node in a Ray task graph.
+
+    A node has a type (e.g., TaskNode), data (e.g., function options and body),
+    arguments (Python values, DAGNodes, and DAGNodes nested within Python
+    argument values) and options (Ray API .options() used for function, class
+    or class method)
+    """
+
+    def __init__(
+        self,
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+        options: Dict[str, Any],
+        kwargs_to_resolve: Optional[Dict[str, Any]] = None,
+    ):
+        # Bound node arguments, ex: func_or_class.bind(1)
+        self._bound_args: Tuple[Any] = [] if args is None else args
+        # Bound node keyword arguments, ex: func_or_class.bind(a=1)
+        self._bound_kwargs: Dict[str, Any] = {} if kwargs is None else kwargs
+        # Bound node options arguments, ex: func_or_class.options(num_cpus=2)
+        self._bound_options: Dict[str, Any] = {} if options is None else options
+        # Bound kwargs to resolve that's specific to subclass implementation
+        self._bound_kwargs_to_resolve: Optional[Dict[str, Any]] = kwargs_to_resolve
+        # UUID that is not changed over copies of this node.
+        self._stable_uuid = uuid.uuid4().hex
+
+    def get_args(self) -> Tuple[Any]:
+        """Return the tuple of arguments for this node."""
+
+        return self._bound_args
+
+    def get_kwargs(self) -> Dict[str, Any]:
+        """Return the dict of keyword arguments for this node."""
+
+        return self._bound_kwargs.copy()
+
+    def get_options(self) -> Optional[Dict[str, Any]]:
+        """Return the dict of options arguments for this node."""
+
+        return self._bound_options.copy()
+
+    def execute(self) -> Union[ray.ObjectRef, ray.actor.ActorHandle]:
```

Comment:
this signature needs to be `def execute(self, *args, **kwargs)` to allow user to pass in runtime requests to execute rather than bound args. It's WIP in a follow-up PR.